### ISSUE TITLE
Model Pruning, Knowledge Distillation or Model Quantization

### ISSUE THREAD

**Issue by hjm1990818** [AlexeyAB/darknet#4053]
Dear @AlexeyAB ,

I design a network and train it on my custom dataset by using darknet framework.

My network weight is about 280MB. I want to use Model compression technology 
(Model Pruning, Knowledge Distillation or Model Quantization) to decrease the memory usage of my network weight, but I don't know how to do it.

I read the work named：Learning Efficient Convolutional Networks through Network Slimming
Althogh the source code of that work can be downloaded through Github. However, the core code of its model compression method is not open in public.

Could you please give me some help to decrease the memory usage of my network weight?

Looking forward to hearing from you, thank you very much in advance.

With best regards,
Ming

**Comment by hjm1990818** [comment]
> Dear @AlexeyAB ,
> 
> I design a network and train it on my custom dataset by using darknet framework.
> 
> My network weight is about 280MB. I want to use Model compression technology
> (Model Pruning, Knowledge Distillation or Model Quantization) to decrease the memory usage of my network weight, but I don't know how to do it.
> 
> I read the work named：Learning Efficient Convolutional Networks through Network Slimming
> Althogh the source code of that work can be downloaded through Github. However, the core code of its model compression method is not open in public.
> 
> Could you please give me some help to decrease the memory usage of my network weight?
> 
> Looking forward to hearing from you, thank you very much in advance.
> 
> With best regards,
> Ming

I am so sorry the paper I read is: SlimYOLOv3: Narrower, Faster and Better for Real-Time UAV Application

Could you please give me some help to decrease the memory usage of my network weight?

@AlexeyAB

**Comment by AlexeyAB** [comment]
@hjm1990818 Hi,

Read this thread: https://github.com/AlexeyAB/darknet/issues/3732

Model Pruning:
It seems that you just can reduce `filters=` /2 or /4 in the middle layers in cfg-file and train from scratch without pre-trained weights-file.

**Comment by hjm1990818** [comment]
> @hjm1990818 Hi,
> 
> Read this thread: #3732
> 
> Or instead of Model Pruning:
> It seems that you just can reduce `filters=` /2 or /4 in the middle layers in cfg-file and train from scratch without pre-trained weights-file.

Dear Dr. AlexeyAB,

Thank you very much for your response.

My network have 30 convolutional layers. As you told me that I just can reduce filters= /2 or /4 in the middle layers in cfg-file and train from scratch without pre-trained weights-file.

I will follow your advice and intend to do this as follows:

**1:** In my original cfg-file, the 15th conv-layer is defined as below:
                                       [convolutional]
                                       batch_normalize=1
                                       filters=64
                                       size=7
                                       stride=2
                                       pad=1
                                       activation=leaky
           Now, I should change it into:
                                       [convolutional]
                                       batch_normalize=1
                                       filters=32
                                       size=7
                                       stride=2
                                       pad=1
                                       activation=leaky
          After that, the 16th, 17th,...., 23th conv-layer do the same operation

**2.** I should train the new cfg-file without pre-trained weights-file.

Am I right to follow the above two steps ? Any mistakes? 

Looking forward to hearing from you, thanks in advance.

With best regards,
Ming

**Comment by hjm1990818** [comment]
> @hjm1990818 Hi,
> 
> Read this thread: #3732
> 
> Or instead of Model Pruning:
> It seems that you just can reduce `filters=` /2 or /4 in the middle layers in cfg-file and train from scratch without pre-trained weights-file.

Dear Dr. @AlexeyAB,

Thank you very much for your response.

My network have 30 convolutional layers. As you told me that I just can reduce filters= /2 or /4 in the middle layers in cfg-file and train from scratch without pre-trained weights-file.

I will follow your advice and intend to do this as follows:

**1:** In my original cfg-file, the 15th conv-layer is defined as below:
[convolutional]
batch_normalize=1
filters=64
size=7
stride=2
pad=1
activation=leaky
Now, I should change it into:
[convolutional]
batch_normalize=1
filters=32
size=7
stride=2
pad=1
activation=leaky
After that, the 16th, 17th,...., 23th conv-layer do the same operation

**2.** I should train the new cfg-file without pre-trained weights-file.

Am I right to follow the above two steps ? Any mistakes?

Looking forward to hearing from you, thanks in advance.

With best regards,
Ming

**Comment by LukeAI** [comment]
@hjm1990818 probably the easiest thing would be to take one of the cfg files from the github of the author of slimYolo https://github.com/PengyiZhang/SlimYOLOv3
I tried training from scratch with their prune0.5 cfg (using their prune weights convolutional layers as a starting point.) and got ok results - On BDD I got something like 64% mAP with yolov3-spp, and about 58% with prune0.5 model (and maybe 45% with tiny-yolo).
You can also reduce GPU usage by using a smaller input size. If you compile darknet with cudnn_half you will get fp16 quantization but it will increase GPU memory usage, the advantage is it will run quicker.

**Comment by akashAD98** [comment]
@AlexeyAB  @May-forever is there any update on this regarding model pruning
